src/libopenrave/kinbodystatesaver.cpp

```diff
@@ -18,6 +18,121 @@
 
 namespace OpenRAVE {
 
+/// \brief push link to _listNonCollidingLinksWhenGrabbed of grabbed.
+static void _PushLinkToListNonCollidingLinksWhenGrabbed(Grabbed& grabbed,
+                                                        const int linkindex,
+                                                        const KinBody::LinkConstPtr& pLinkRef,
+                                                        const std::vector<KinBody::LinkPtr>& vLinks,
+                                                        const EnvironmentBasePtr& pEnv)
+{
+    if( linkindex < 0 || linkindex >= (int)vLinks.size() ) {
+        RAVELOG_WARN_FORMAT("env=%s, could not restore link '%s' since its index %d is out of range (body num links is %d)", pEnv->GetNameId()%pLinkRef->GetName()%pLinkRef->GetIndex()%vLinks.size());
+    }
+    else {
+        grabbed._listNonCollidingLinksWhenGrabbed.push_back(vLinks.at(linkindex));
+    }
+}
+
+void KinBody::_RestoreGrabbedBodiesFromSavedData(const KinBody& savedBody,
+                                                 const int options,
+                                                 const std::unordered_map<int, GrabbedPtr>& savedGrabbedBodiesByEnvironmentIndex)
+{
+    {
```

Comment:
Why is the entire function doubly-nested?

Reply:
The motivation is same as https://github.com/rdiankov/openrave/pull/1429#discussion_r1767724097.

I originally tried to cleanup the codes, by minimizing the `diff`.
The original code is like this:
```
    if( _options & Save_GrabbedBodies ) {
        xxx
    }
```
And, i just removed this `if` statement, and avoided from other cleaning up to save the diff lines.

However, I now ended up with separating the duplicated code into one function, and thus, we already have huge `diff`.

So, I don't need to stick to the original code, so i'll try to resolve this, and https://github.com/rdiankov/openrave/pull/1429#discussion_r1767724097.

Thanks!

Reply:
fixed via https://github.com/rdiankov/openrave/pull/1429/commits/4970937f95fa69faa5839465e0b3b21be80db0bf